src/platform/storage_winkernel.c

```diff
@@ -506,3 +580,178 @@ CxPlatStorageReadValue(
 
     return Status;
 }
+
+_IRQL_requires_max_(PASSIVE_LEVEL)
+QUIC_STATUS
+CxPlatStorageWriteValue(
+    _In_ CXPLAT_STORAGE* Storage,
+    _In_z_ const char * Name,
+    _In_ CXPLAT_STORAGE_TYPE Type,
+    _In_ uint32_t BufferLength,
+    _In_reads_bytes_(BufferLength)
+        const uint8_t * Buffer
+    )
+{
+    QUIC_STATUS Status;
+    PUNICODE_STRING NameString = NULL;
+
+    Status = CxPlatConvertUtf8ToUnicode(Name, QUIC_POOL_PLATFORM_TMP_ALLOC, &NameString);
+    if (QUIC_FAILED(Status)) {
+        return Status;
+    }
+
+    Status =
+        ZwSetValueKey(
+            Storage->RegKey,
+            NameString,
+            0,
+            (ULONG)Type,
+            (PVOID)Buffer,
+            BufferLength);
+
+    CXPLAT_FREE(NameString, QUIC_POOL_PLATFORM_TMP_ALLOC);
+
+    return Status;
+}
+
+_IRQL_requires_max_(PASSIVE_LEVEL)
+QUIC_STATUS
+CxPlatStorageDeleteValue(
+    _In_ CXPLAT_STORAGE* Storage,
+    _In_z_ const char * Name
+    )
+{
+    QUIC_STATUS Status;
+    PUNICODE_STRING NameString = NULL;
+
+    Status = CxPlatConvertUtf8ToUnicode(Name, QUIC_POOL_PLATFORM_TMP_ALLOC, &NameString);
+    if (QUIC_FAILED(Status)) {
+        return Status;
+    }
+
+    Status =
+        ZwDeleteValueKey(
+            Storage->RegKey,
+            NameString);
+
+    CXPLAT_FREE(NameString, QUIC_POOL_PLATFORM_TMP_ALLOC);
+
+    return Status;
+}
+
+_IRQL_requires_max_(PASSIVE_LEVEL)
+QUIC_STATUS
+CxPlatStorageClear(
+    _In_ CXPLAT_STORAGE* Storage
+    )
+{
+    QUIC_STATUS Status = QUIC_STATUS_SUCCESS;
+    ULONG InfoLength = 0;
+    ULONG AllocatedLength = 0;
+    PKEY_VALUE_BASIC_INFORMATION Info = NULL;
+
+    //
+    // Iterate through all values and delete them
+    // We always use index 0 because deletion shifts the remaining values
+    //
+    while (TRUE) {
+        //
+        // First, query the required buffer size for the value name
+        //
+        Status = ZwEnumerateValueKey(
+            Storage->RegKey,
+            0, // Always use index 0 since we delete as we go
+            KeyValueBasicInformation,
+            NULL,
```

Comment:
nit: Instead of always doing a double-call, you can provide whatever buffer we have here.
If it is big enough, the call will succeed on the first try, and we don't need to try a second time.

We can also start by pre-allocating a reasonable size (30 characters + offset), this way, we avoid a bad pattern where we keep reallocating because the name of the next key keeps being 1 character longer than the previous one.